### PR TITLE
Zi/refine certdb

### PR DESCRIPTION
**Pull request by lziest**
address #503 and added a little postgres schema refinement.

### CODE REVIEW

**Comment by lziest** [comment]
@jacobhaven please review. :)

File: certdb/pg/migrations/001_CreateCertificates.sql [hunk 1]
```diff
@@ -3,16 +3,26 @@
 
 CREATE TABLE certificates (
   id                serial,
-  serial            bytea NOT NULL PRIMARY KEY,
+  serial            bytea NOT NULL UNIQUE,
```
Comment (jacobhaven) [inline certdb/pg/migrations/001_CreateCertificates.sql]: Should rename this to `serial_num` to not conflict with Postgres type

File: certdb/pg/migrations/001_CreateCertificates.sql [hunk 2]
```diff
@@ -3,16 +3,26 @@
 
 CREATE TABLE certificates (
   id                serial,
```
Comment (jacobhaven) [inline certdb/pg/migrations/001_CreateCertificates.sql]: I'm not sure need these `id` fields.

File: certdb/pg/migrations/001_CreateCertificates.sql [hunk 3]
```diff
@@ -3,16 +3,26 @@
...
   revoked_at        timestamptz,
-  pem               bytea NOT NULL
+  pem               bytea NOT NULL,
+  PRIMARY KEY(id, serial)
 );
 
+CREATE TABLE ocsp_responses (
+  id                       serial PRIMARY KEY,
+  serial                   bytea NOT NULL,
+  authority_key_identifier bytea NOT NULL,
```
Comment (jacobhaven) [inline certdb/pg/migrations/001_CreateCertificates.sql]: I would move `authority_key_identifier` to the `certificates` table and have

```SQL
PRIMARY KEY(serial, authority_key_identifier)
```

File: certdb/sqlite/migrations/002_CreateOCSPResponses.sql [hunk 4]
```diff
@@ -2,9 +2,10 @@
 -- SQL in section 'Up' is executed when this migration is applied
```
Comment (jacobhaven) [inline certdb/sqlite/migrations/002_CreateOCSPResponses.sql]: This is not the same as the Postgres table. Should probably have them match :P

**Comment by lziest** [comment]
@jacobhaven @drcapulet is it OK I merge this by EOW?

**Comment by drcapulet** [comment]
Overall LGTM, though I do find it odd we return slices in `GetCertificate` & `GetOCSP` while the method names sound singular and in the DB we're uniquely identifying both certificates + OCSP responses by (serial, aki).

**Comment by lziest** [comment]
I understand. We can change that later. I want to settle the DB schema now, interface can definitely evolve. I want to do "insertCertificate(cert *x509.Certificate)" as well so we don't need to remember hex encoding the hash value. And if you are really annoyed, let's do it in your PR. :)

File: certdb/pg/migrations/001_CreateCertificates.sql [hunk 5]
```diff
@@ -2,17 +2,27 @@
...
 CREATE TABLE certificates (
-  id                serial,
-  serial            bytea NOT NULL PRIMARY KEY,
-  ca_label          bytea NOT NULL,
-  status            bytea NOT NULL,
-  reason            int,
-  expiry            timestamptz,
-  revoked_at        timestamptz,
-  pem               bytea NOT NULL
+  serial_number            bytea NOT NULL UNIQUE,
```
Comment (drcapulet) [inline certdb/pg/migrations/001_CreateCertificates.sql]: Do we need the `UNIQUE` constraint here since the `PRIMARY KEY` covers it?

Comment (lziest) [inline certdb/pg/migrations/001_CreateCertificates.sql]: good point, we can remove it

**Comment by lziest** [comment]
I am happy to merge now

**Comment by kisom** [comment]
LGTM.

**Pull request merged**